server/manager/src/main/java/org/apache/accumulo/manager/TabletGroupWatcher.java

```diff
@@ -669,8 +598,97 @@ private void updateMergeState(Map<TableId,MergeStats> mergeStatsCache) {
+      // If needsFencingForDeletion() is true then this tablet was previously fenced
+      // so we need to keep it and not delete the tablet. Skip to the next highest
+      // tablet after this one if it exists
+      if (needsFencingForDeletion(info, firstTablet)) {
+        Key nextExtent = new Key(deletionStartRow).followingKey(PartialKey.ROW);
+        deletionStartRow = getHighTablet(new KeyExtent(info.getExtent().tableId(),
+            nextExtent.getRow(), info.getExtent().prevEndRow())).endRow();
+        Manager.log.trace("First tablet in deletion range updated to contain startRow {}",
+            deletionStartRow);
+      }
+    }
+    return deletionStartRow;
+  }
+
+  private Text getDeletionEndRow(MergeInfo info) {
+    Text deletionEndRow = info.getExtent().endRow();
+    if (deletionEndRow != null) {
+      Manager.log.trace("Finding tablet that contains the end row {} for deletion", deletionEndRow);
+      final KeyExtent last =
+          loadTabletMetadata(info.getExtent().tableId(), deletionEndRow, ColumnType.PREV_ROW)
+              .map(TabletMetadata::getExtent).orElse(null);
+      // If needsFencingForDeletion() is true then this tablet was previously fenced
+      // so we need to keep it and not delete the tablet. Find the previous tablet
+      // if it exists to stop deletion on
+      if (needsFencingForDeletion(info, last)) {
+        deletionEndRow = last.prevEndRow();
+        Manager.log.debug("Last tablet in deletion range updated to contain endRow {}",
+            deletionEndRow);
+      }
+    }
+    return deletionEndRow;
+  }
+
+  private boolean hasTabletsToDelete(MergeInfo info, Text deletionStartRow, Text deletionEndRow) {
+    // If the endRow of the deletion range is not null, but the actual deletionEndRow is null then
+    // this means
+    // the first tablet was fenced so there are no tablets to delete and we can return
+    if (info.getExtent().endRow() != null && deletionEndRow == null) {
+      return false;
+      // If the prevEndRow of the deletion range is not null, but the actual deletionStartRow is
+      // null then this means
+      // the last tablet was fenced so there are no tablets to delete and we can return
+    } else if (info.getExtent().prevEndRow() != null && deletionStartRow == null) {
+      return false;
+      // Make sure detected deletionEndRow is after deletionStartRow. If they are the same tablet or
+      // before then nothing to delete and we can return
+      // This would happen in some situations such as a deletion range fully contained in 1 tablet
+    } else if (deletionEndRow != null && deletionStartRow != null
+        && new Key(deletionEndRow).compareRow(deletionStartRow) <= 0) {
+      return false;
+    }
+
+    return true;
+  }
+
   private void deleteTablets(MergeInfo info) throws AccumuloException {
-    KeyExtent extent = info.getExtent();
+    // Before deleting actual tablets we need to fence the files in tablets
```

Comment:
Still have not completed reviewing this.  Looking through what I have so far I am wondering if it would simplify things to merge the new and old code into a single scan of the metadata table that does everything.  This single scan could keep state as it goes and avoid having to recompute things. Was thinknig of something like the following.

```java
  private void deleteTablets(MergeInfo info) throws Exception {
    Preconditions.checkArgument(info.isDelete());
    var deleteRange = info.getExtent();

    // TODO hard coded table name
    try (var tablets = manager.getContext().getAmple().readTablets().forTable(deleteRange.tableId())
        .overlapping(deleteRange.prevEndRow(), false, null).checkConsistency().saveKeyValues()
        .build(); var writer = manager.getContext().createBatchWriter("accumulo.metadata")) {
      KeyExtent firstTabletDeleted = null;
      for (var tablet : tablets) {
        var tabletExtent = tablet.getExtent();
        if (deleteRange.contains(tabletExtent)) {
          // delete tablet because its completely contained in delete range
          deleteTablet(writer, tablet);

          if (firstTabletDeleted == null) {
            firstTabletDeleted = tabletExtent;
          }
        } else if (deleteRange.overlaps(tabletExtent)) {
          // this could be the first tablet, last tablet, or both first and last tablet
          fenceTablet(writer, deleteRange, tablet);

          if (deleteRange.endRow() != null && tabletExtent.contains(deleteRange.endRow())) {
            // this is the last tablet
            if (firstTabletDeleted != null) {
              // TODO could fence set prevend row in same mutation that fences
              // some tablets were deleted so need to reset this tablets prev end row
              setPrevEndRow(writer, tablet, firstTabletDeleted.prevEndRow());
            }
            // saw the last tablet and fixed it up, so done
            break;
          }
        } else {
          // this tablet is disjoint with the deletion range AND we did not see a last tablet that
          // partially overlaps, so need to fence it and set its prev end row
          fenceTablet(writer, deleteRange, tablet);

          // adjust the tablets prev end row
          setPrevEndRow(writer, tablet, firstTabletDeleted.prevEndRow());
        }
      }
    }
  }
```

While writing this I realized this code needs to be idempotent and the above code is not idempotent, but I think could be made so with some slight modifications.  Do you know if the changes you have made so far are idempotent?

Reply:
I can take a look at implementing something like this where we change the code to just do one pass of the metadata vs the current version which does 2 passes (first updates/fences and then the deletion of the tablets). Instead of deleting tablets another option is to just drop the files and run the merge command after as a follow on. I can explore these various things in a separate branch and follow on since they are big changes.

I am not sure if the current version is idempotent, I would need to look at it and think about it in terms of what happens in failure scenarios.